Standards/scs-0215-v1-robustness-features.md

```diff
@@ -0,0 +1,252 @@
+in order to protect the server against outages, deceleration or malfunctions due to an
+overload of requests.
+In order to do so, at least the following parameters SHOULD be set on a Kubernetes cluster:
+
+* max-requests-inflight
+* max-mutating-requests-inflight
+* min-request-timeout
+
+Values for these flags/parameters MUST be adapted to the needs of the environment and
+the expected load.
+
+A cluster MUST also activate and configure a Ratelimit admission controller.
+This requires an `EventRateLimit` resource to be deployed on the Kubernetes cluster.
+The following settings are RECOMMENDED for a cluster-wide deployment, but more
+fine-grained rate limiting can also be applied, if this is necessary.
+
+```yaml
+kind: Configuration
+apiVersion: eventratelimit.admission.k8s.io/v1alpha1
+limits:
+- burst: 20000
+  qps: 5000
+  type: Server
+```
+
+It is also RECOMMENDED to activate the Kubernetes API priority and fairness feature,
+which also uses the aforementioned cluster parameters to better queue, schedule and
+prioritize incoming requests.
+
+### etcd compaction
+
+etcd MUST be cleaned up regularly, so that it functions correctly and doesn't take
+up too much space, which happens because of its increase of the keyspace.
+
+To compact the etcd keyspace, the following flags/parameters MUST be set for etcd:
+
+* auto-compaction-mode = periodic
+* auto-compaction-retention = 8h
+
+### etcd backup
+
+An etcd cluster MUST be backed up regularly. It is RECOMMENDED to adapt
+a strategy of decreasing backups over longer time periods, e.g. keeping snapshots every
+10 minutes for the last 120 minutes, then one hourly for 1 day, then one daily for 2 weeks,
+then one weekly for 3 months, then one monthly for 2 years, and after that a yearly backup.
+At the very least, a backup MUST be done once a week.
+These numbers can be adapted to the security setup and concerns like storage or network
+usage. It is also RECOMMENDED to encrypt the backups in order to secure them further.
+How this is done is up to the operator.
+
+### CA expiration avoidance
+
+It should be avoided, that certificates expire either on the whole cluster or for single components.
+To avoid this scenario, certificates SHOULD be rotated regularly; in the
+case of SCS, we REQUIRE at least a yearly certificate rotation.
+To achieve a complete certificate rotation, the parameters `serverTLSBootstrap` and `rotateCertificates` MUST be set.
+
+The certificates can be rotated by either updating the Kubernetes cluster, which automatically
+renews certificates, or by manually renewing them with the command
+
+```bash
+kubeadm certs renew all
+```
+
+After this, new CSRs MUST be approved manually, normally done with
+
+```bash
+kubectl get csr
+kubectl certificate approve <CSR>
+```
+
+or be approved with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).
```

Comment:
I changed a few sentences in the document and especially in the `Decision` section in order to better reflect a few of the concerns you had. You're right that `kubeadm` was a bit to specific, so I only mentioned it as an example in the `Design Considerations`.

About the CSR part, I'm not really sure. Yes they don't HAVE to be used, but https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs reads like its a recommended way, especially if some external tools should be used with the cluster (like monitoring solutions).
But I'm really not an expert on this.

Reply:
@joshmue Is this fine for now or are there any other problems with this?

Reply:
> You're right that kubeadm was a bit to specific, so I only mentioned it as an example in the Design Considerations.

:+1: 

> About the CSR part, I'm not really sure. [...] But I'm really not an expert on this.

I did not dig into that to much right now, so I'm not totally sure whether this should be mandated here, either (I'd tend not to).
If this is to be included and merged, someone should be sure about that, though.

> Yes they don't HAVE to be used, but https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs reads like its a recommended way, especially if some external tools should be used with the cluster (like monitoring solutions).

Again, that is kubeadm-centric documentation, so not necessarily relevant for other distributions.

Other than that, the current status has some SHOULD/can/MUST ambiguity problems:

> To avoid this scenario, certificates SHOULD be rotated regularly; in the
> case of SCS, we REQUIRE at least a yearly certificate rotation.

That could be understood two ways: Either someone could decide to be non-compliant with that SHOULD, hence making the REQUIRE-ment ineffective (That again would mandate *how* something is done, while the *what* is optional - pretty much the opposite of what standards usually do). Or someone could read the REQUIRE as actual requirement, making the SHOULD an effective MUST.

> The certificates can be rotated by either updating the Kubernetes cluster, which automatically
> renews certificates, or by manually renewing them. How this is done is dependent on the used K8s cluster.
> 
> After this, new CSRs MUST be approved manually or with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).

Kind of the same problem as above.

> It is also RECOMMENDED to renew the CA regularly to avoid an expiration of the CA.

:+1: 

> This standard doesn't set a timeline for this, since it is dependent on the CA.

Is this intended to say that the CA may be out of the cluster operator's control? That's not really clear.

Reply:
> > Yes they don't HAVE to be used, but https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs reads like its a recommended way, especially if some external tools should be used with the cluster (like monitoring solutions).
>
> Again, that is kubeadm-centric documentation, so not necessarily relevant for other distributions.

Yes this documentation uses `kubeadm`, but I would argue that the CSR part is not-centered especially on `kubeadm` and can be applied to other K8s cluster "types" aswell.

> > To avoid this scenario, certificates SHOULD be rotated regularly; in the
> > case of SCS, we REQUIRE at least a yearly certificate rotation.
>
> That could be understood two ways: Either someone could decide to be non-compliant with that SHOULD, 
> hence making the REQUIRE-ment ineffective (That again would mandate *how* something is done, while the 
> *what* is optional - pretty much the opposite of what standards usually do). Or someone could read the 
> REQUIRE as actual requirement, making the SHOULD an effective MUST.
>
> > The certificates can be rotated by either updating the Kubernetes cluster, which automatically
> > renews certificates, or by manually renewing them. How this is done is dependent on the used K8s cluster.
> > 
> > After this, new CSRs MUST be approved manually or with a third-party controller, e.g. the [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver).
>
> Kind of the same problem as above.

I agree with the first part (and I fixed that), but the second part doesn't have the same problem.
The whole part is about certificate rotation (meaning its continuous), so the part with the `can` is refering to the certificate rotation part, which is written as a `MUST` now (`SHOULD` before, so that would be contradictive).
Also `can` isn't an RFC keyword (https://www.rfc-editor.org/rfc/rfc2119), so it's not defining anything, just explaining a process from this.

Reply:
Updated with https://github.com/SovereignCloudStack/standards/pull/361/commits/84fcef3ff4670d82ac28ab5cbf85caf515678a01

Reply:
> > > Yes they don't HAVE to be used, but https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs reads like its a recommended way, especially if some external tools should be used with the cluster (like monitoring solutions).
> >
> > Again, that is kubeadm-centric documentation, so not necessarily relevant for other distributions.
> 
> Yes this documentation uses `kubeadm`, but I would argue that the CSR part is not-centered especially on `kubeadm` and can be applied to other K8s cluster "types" aswell.

I did not really use `kubeadm` all that much for a few years now. But looking at its [docs](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#renew-certificates-with-the-kubernetes-certificates-api) (mind the `Caution` paragraph), handling certificates this way seems to be the exception even for `kubeadm` setups, nevermind other distributions.

> I agree with the first part (and I fixed that)

:+1: 

> The whole part is about certificate rotation (meaning its continuous), so the part with the can is refering to the certificate rotation part, which is written as a MUST now (SHOULD before, so that would be contradictive).
> Also can isn't an RFC keyword (https://www.rfc-editor.org/rfc/rfc2119), so it's not defining anything, just explaining a process from this.

Still, this MUST refers to `how` certificates should be approved. In some setups, there probably is not even a concept for approval present.

And if there is, it is buried so low that it is unpractical to check for.
For example: Could [k3s certificate rotation](https://docs.k3s.io/cli/certificate) ever be compliant to this standard's MUST? That is cumbersome to (dis-)prove and does not add much value over the simple requirement of "Certificates should be rotated regularly".

Reply:
> > The whole part is about certificate rotation (meaning its continuous), so the part with the can is refering to the > > certificate rotation part, which is written as a MUST now (SHOULD before, so that would be contradictive).
> >Also can isn't an RFC keyword (https://www.rfc-editor.org/rfc/rfc2119), so it's not defining anything, just explaining a process from this.
>
> Still, this MUST refers to how certificates should be approved. In some setups, there probably is not even a
> concept for approval present.
>
> And if there is, it is buried so low that it is unpractical to check for.
> For example: Could [k3s certificate rotation](https://docs.k3s.io/cli/certificate) ever be compliant to this 
> standard's MUST? That is cumbersome to (dis-)prove and does not add much value over the simple requirement
> of "Certificates should be rotated regularly".

Removed the sentence in question.